File: backend/scoring/moves.py
"""Server-side port of Stage 4 - move segmentation, static/dynamic
classification, per-move metrics. Mirrors frontend/src/pipeline/moves.ts.
See PIPELINE.md - Stage 4.

Settled Decision: static vs dynamic by a velocity/acceleration threshold on the
smoothed CoM trajectory, NOT a trained classifier in V1. Static moves are scored
on CoM-to-target distance at the reach; dynamic moves on landing control only
(never mid-move distance - a lunge requires the CoM to be far). No "was the dyno
necessary" judgment (deferred).
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Literal

from scoring.cog import LEFT_WRIST, RIGHT_WRIST, Frame
from scoring.smoothing import SmoothedTrajectory, magnitude

MoveType = Literal["static", "dynamic"]


@dataclass(frozen=True)
class MoveParams:
    arrival_radius: float = 0.05
    dynamic_accel_threshold: float = 3.0
    landing_window_seconds: float = 0.3
    confidence_threshold: float = 0.5


MOVE_PARAMS = MoveParams()


@dataclass(frozen=True)
class Hold:
    sequence_index: int
    frame_x: float
    frame_y: float
    id: str | None = None  # server-side hold id, when known


@dataclass(frozen=True)
class Move:
    move_index: int
    target_hold_sequence_index: int
    target_hold_id: str | None
    move_type: MoveType
    cog_distance: float | None
    landing_control: float | None
    confidence: float
    low_confidence: bool
    start_frame: int
    end_frame: int


def _distance(ax: float, ay: float, bx: float, by: float) -> float:
    return math.hypot(ax - bx, ay - by)


def _wrist_distance_to_hold(frame: Frame, hold: Hold) -> float:
    lw = frame[LEFT_WRIST]
    rw = frame[RIGHT_WRIST]
    return min(
        _distance(lw.x, lw.y, hold.frame_x, hold.frame_y),
        _distance(rw.x, rw.y, hold.frame_x, hold.frame_y),
    )


def _find_arrival_frame(frames: list[Frame], hold: Hold, from_frame: int) -> int:
    """First frame from `from_frame` where a wrist is within the arrival radius;
    else the frame where a wrist gets closest (so segmentation still produces a
    window even if the hold was never cleanly reached)."""
    best_frame = from_frame
    best_dist = math.inf
    for i in range(from_frame, len(frames)):
        d = _wrist_distance_to_hold(frames[i], hold)
        if d < MOVE_PARAMS.arrival_radius:
            return i
        if d < best_dist:
            best_dist = d
            best_frame = i
    return best_frame


def classify_moves(
    frames: list[Frame], trajectory: SmoothedTrajectory, holds: list[Hold]
) -> list[Move]:
    if len(frames) != len(trajectory.points):
        raise ValueError(
            f"frames ({len(frames)}) and trajectory ({len(trajectory.points)}) lengths differ"
        )
    if not holds:
        raise ValueError("cannot segment moves with zero annotated holds")

    ordered = sorted(holds, key=lambda h: h.sequence_index)
    moves: list[Move] = []
    window_start = _find_arrival_frame(frames, ordered[0], 0)

    for h in range(1, len(ordered)):
        hold = ordered[h]
        search_from = min(window_start + 1, len(frames) - 1)
        arrival = _find_arrival_frame(frames, hold, search_from)
        start_frame = window_start
        end_frame = max(arrival, start_frame)
        moves.append(_score_move(len(moves), hold, start_frame, end_frame, trajectory))
        window_start = end_frame

    return moves


def _score_move(
    move_index: int, hold: Hold, start_frame: int, end_frame: int, trajectory: SmoothedTrajectory
) -> Move:
    points = trajectory.points
    acceleration = trajectory.acceleration
    frame_rate = trajectory.frame_rate

    peak_accel = 0.0
    confidence_sum = 0.0
    for i in range(start_frame, end_frame + 1):
        peak_accel = max(peak_accel, magnitude(acceleration[i]))
        confidence_sum += points[i].confidence
    confidence = confidence_sum / (end_frame - start_frame + 1)
    move_type: MoveType = (
        "dynamic" if peak_accel > MOVE_PARAMS.dynamic_accel_threshold else "static"
    )

    cog_distance: float | None = None
    landing_control: float | None = None
    if move_type == "static":
        cog = points[end_frame]
        cog_distance = _distance(cog.x, cog.y, hold.frame_x, hold.frame_y)
    else:
        landing_frames = max(1, round(MOVE_PARAMS.landing_window_seconds * frame_rate))
        landing_end = min(end_frame + landing_frames, len(points) - 1)
        speed_sum = 0.0
        count = 0
        for i in range(end_frame, landing_end + 1):
            speed_sum += magnitude(trajectory.velocity[i])
            count += 1
        landing_control = speed_sum / count

    return Move(
        move_index=move_index,
        target_hold_sequence_index=hold.sequence_index,
        target_hold_id=hold.id,
        move_type=move_type,
        cog_distance=cog_distance,
        landing_control=landing_control,
        confidence=confidence,
        low_confidence=confidence < MOVE_PARAMS.confidence_threshold,
        start_frame=start_frame,
        end_frame=end_frame,
    )

File: frontend/src/pipeline/moves.ts
// Stage 4 - move segmentation, static/dynamic classification, per-move metrics
// (PIPELINE.md). Settled Decision: static vs dynamic by a velocity/acceleration
// threshold on the smoothed CoM trajectory, NOT a trained classifier in V1.
//
// Segmentation ties each move window to the interval ending when a hand reaches
// the next annotated hold (tap order = intended sequence). Deliberately simple
// and explainable.

import { LANDMARK } from './cog';
import { magnitude } from './smoothing';
import type { FrameLandmarks, Hold, Move, SmoothedTrajectory } from './types';

/**
 * Tunable thresholds, named so they can be re-tuned against the fixture clip.
 * All distances are in normalized frame units; time-derived values use real
 * seconds via the frame rate.
 */
export const MOVE_PARAMS = {
  /** A wrist within this distance of a hold counts as arrived at it. */
  arrivalRadius: 0.05,
  /**
   * Peak CoM acceleration (normalized units/s^2) above which a move is dynamic.
   * A controlled static reach barely accelerates the CoM; a lunge/dyno produces
   * a sharp spike an order of magnitude larger. Re-tune on real footage.
   */
  dynamicAccelThreshold: 3.0,
  /** Seconds after the catch over which landing control is measured. */
  landingWindowSeconds: 0.3,
  /** Mean pose confidence below which a move is flagged low-confidence. */
  confidenceThreshold: 0.5,
} as const;

export type MoveParams = typeof MOVE_PARAMS;

function distance(ax: number, ay: number, bx: number, by: number): number {
  return Math.hypot(ax - bx, ay - by);
}

/** Nearest-wrist distance to a hold for one frame. */
function wristDistanceToHold(frame: FrameLandmarks, hold: Hold): number {
  const lw = frame[LANDMARK.LEFT_WRIST];
  const rw = frame[LANDMARK.RIGHT_WRIST];
  return Math.min(
    distance(lw.x, lw.y, hold.frameX, hold.frameY),
    distance(rw.x, rw.y, hold.frameX, hold.frameY),
  );
}

/**
 * Find the frame at which a hand arrives at the hold, searching from
 * `fromFrame`: the first frame where a wrist is within the arrival radius. If
 * no frame qualifies (hold missed, occlusion), fall back to the frame where a
 * wrist gets closest, so segmentation still produces a window.
 */
function findArrivalFrame(frames: FrameLandmarks[], hold: Hold, fromFrame: number): number {
  let bestFrame = fromFrame;
  let bestDist = Number.POSITIVE_INFINITY;
  for (let i = fromFrame; i < frames.length; i++) {
    const d = wristDistanceToHold(frames[i], hold);
    if (d < MOVE_PARAMS.arrivalRadius) return i;
    if (d < bestDist) {
      bestDist = d;
      bestFrame = i;
    }
  }
  return bestFrame;
}

/**
 * Segment the climb into moves and score each one.
 *
 * - Holds must be the user's annotated set; tap order = intended sequence.
 * - The first hold in sequence is treated as the starting hold; every
 *   subsequent hold defines one move (the reach toward it).
 * - Static moves are scored on CoM-to-target distance at the reach.
 * - Dynamic moves are scored on landing control ONLY - mid-move CoM distance is
 *   never scored, because a lunge requires the CoM to be far from the target.
 *   No "was this dyno necessary" judgment (deferred).
 */
export function classifyMoves(
  frames: FrameLandmarks[],
  trajectory: SmoothedTrajectory,
  holds: Hold[],
): Move[] {
  if (frames.length !== trajectory.points.length) {
    throw new Error(
      `frames (${frames.length}) and trajectory (${trajectory.points.length}) lengths differ`,
    );
  }
  if (holds.length === 0) {
    throw new Error('cannot segment moves with zero annotated holds');
  }

  const ordered = [...holds].sort((a, b) => a.sequenceIndex - b.sequenceIndex);
  const moves: Move[] = [];
  let windowStart = findArrivalFrame(frames, ordered[0], 0);

  for (let h = 1; h < ordered.length; h++) {
    const hold = ordered[h];
    const searchFrom = Math.min(windowStart + 1, frames.length - 1);
    const arrival = findArrivalFrame(frames, hold, searchFrom);
    const startFrame = windowStart;
    const endFrame = Math.max(arrival, startFrame);

    moves.push(scoreMove(moves.length, hold, startFrame, endFrame, trajectory));
    windowStart = endFrame;
  }

  return moves;
}

function scoreMove(
  moveIndex: number,
  hold: Hold,
  startFrame: number,
  endFrame: number,
  trajectory: SmoothedTrajectory,
): Move {
  const { points, acceleration, frameRate } = trajectory;

  let peakAccel = 0;
  let confidenceSum = 0;
  for (let i = startFrame; i <= endFrame; i++) {
    peakAccel = Math.max(peakAccel, magnitude(acceleration[i]));
    confidenceSum += points[i].confidence;
  }
  const confidence = confidenceSum / (endFrame - startFrame + 1);
  const moveType = peakAccel > MOVE_PARAMS.dynamicAccelThreshold ? 'dynamic' : 'static';

  let cogDistance: number | null = null;
  let landingControl: number | null = null;
  if (moveType === 'static') {
    // Core metric: how close the CoM was to the target at the moment of the reach.
    const cog = points[endFrame];
    cogDistance = distance(cog.x, cog.y, hold.frameX, hold.frameY);
  } else {
    // Landing control: mean CoM speed over the window just after the catch.
    // A settled CoM means the dyno was caught in control.
    const landingFrames = Math.max(1, Math.round(MOVE_PARAMS.landingWindowSeconds * frameRate));
    const landingEnd = Math.min(endFrame + landingFrames, points.length - 1);
    let speedSum = 0;
    let count = 0;
    for (let i = endFrame; i <= landingEnd; i++) {
      speedSum += magnitude(trajectory.velocity[i]);
      count++;
    }
    landingControl = speedSum / count;
  }

  return {
    moveIndex,
    targetHoldId: hold.id,
    targetHoldSequenceIndex: hold.sequenceIndex,
    moveType,
    cogDistance,
    landingControl,
    confidence,
    lowConfidence: confidence < MOVE_PARAMS.confidenceThreshold,
    startFrame,
    endFrame,
  };
}
